Purchases: Add option to remove expiring purchases

drewblaisdell opened the pull request:
Fixes #1402 by adding an option to `ManagePurchase` that opens a dialog to remove an expiring purchase.

![screen shot 2016-01-19 at 4 59 25 pm](https://cloud.githubusercontent.com/assets/1130674/12436941/559b0234-bece-11e5-87cf-1ed0aef99559.png)

**Testing**
- Point your hosts file at your sandbox.
- Apply this patch: D906-code
- Visit `/purchases/:site/:purchase_id` for a site/purchase that has auto-renew disabled and is outside the refund period.
- You should see an option to 'Remove [product]'. Clicking it will open the dialog above.
- Click 'Remove' and assert that the buttons are disabled, and that you are eventually redirected to `/purchases` with a notice indicating that the purchase was removed.
- Assert that the purchase that was removed no longer appears in the list below.

- [ ] Product review
- [x] Code review

stephanethomas commented on client/me/purchases/remove-purchase/index.jsx:
```diff
@@ -0,0 +1,110 @@
+
+	getInitialState() {
+		return {
+			isDialogOpen: false
```
Shouldn't we name this property `showDialog` as we did in other components? That's what the [readme file](https://github.com/Automattic/calypso-pre-oss/tree/master/client/components/dialog) suggests in its examples.

drewblaisdell replied:
I'm a fan of prefixing boolean values with "is" to indicate that they are boolean. `showDialog` without context sounds like it actually shows the dialog (like `closeDialog`, which is a method name this file uses). A couple checkout components use `isDialogVisible` for this value, presumably for the same reason.

stephanethomas replied:
I'm this sort of fan as well - I just wanted to stay consistent with the rest of our code. Looks like you updated it to `isDialogVisible`, that's perfect.

drewblaisdell commented:
One outstanding issue here (possibly the only one) is that when the `/me/purchases/:purchase_id/remove` endpoint returns new purchases, these new purchases are added to `PurchasesStore`, but the purchase that was removed remains in memory and still appears on `PurchasesList` until the page is refreshed.

stephanethomas commented on client/me/purchases/manage-purchase/index.jsx:
```diff
@@ -21,6 +22,7 @@ import NoticeAction from 'components/notice/notice-action';
 import { oldShowcaseUrl } from 'lib/themes/helpers';
 import paths from '../paths';
 import PaymentLogo from 'components/payment-logo';
+import RemovePurchase from '../remove-purchase';
```
This should be one line below in order to keep imports sorted.

mikeshelton1503 commented:
For the remove option, let's use the Trash pattern by adding the trash can icon. Also, let's remove the arrow icon from the right since it's not going taking you to a new screen:

![image](https://cloud.githubusercontent.com/assets/12596797/12481127/01d05c4c-c015-11e5-8991-17ca8b07c131.png)

stephanethomas commented:
I've removed some old code, fixed expired purchases that weren't taken into account, and updated the remove link to look as requested.

drewblaisdell commented:
Thanks @stephanethomas. I added https://github.com/Automattic/wp-calypso/commit/da94c396646978cb91b109c816b32c7cd1a6fd7c to remove purchases from the `PurchasesStore` when they are not present in the response from either endpoint that returns purchases. This should be ready for review now.

drewblaisdell commented on client/me/purchases/remove-purchase/index.jsx:
```diff
@@ -0,0 +1,139 @@
+
+		return (
+			<CompactCard className="remove-purchase__card" onClick={ this.openDialog }>
+				<a >
```
@stephanethomas Do you think it is better to use an anchor with no `href` attribute here, or use `cursor: pointer;`?

stephanethomas replied:
The problem with using an `href`attribute in this anchor is that the cursor will only change if you hover the link itself whereas we probably want to change it whenever the cursor enters the card (especially since the `onclick` event applies to the whole card). So `cursor: pointer;` would be needed anyway (@ebinnion or @lezama: that's probably something you want to fix in the `<DeleteUser />` component).

Note I had to inject this anchor in this `CompactCard` because I wasn't able to use its `href` prop in a way which would give me similar results.

stephanethomas replied:
Note I don't mind adding `href="#"` to this anchor since this is pretty much a convention. If not, we should get rid of this unnecessary white space though.

fabianapsimoes commented:
We have a different copy for the dialog when removing domains (https://github.com/Automattic/wp-calypso/issues/1402#issuecomment-170981224):

> Remove [domain name]
> This will remove [domain name] from your account. By removing, you are canceling the domain registration. This may stop you from using it again, even with another service.

Right now, it shows the same copy for all kinds of upgrades:

fabianapsimoes commented:
This doesn't look great in smaller screens if the domain name is a bit long:

![screenshot-calypso localhost 3000 2016-01-22 12-21-38](https://cloud.githubusercontent.com/assets/1248436/12509454/c85ad6f6-c102-11e5-8d12-803bc09ebaf5.png)

fabianapsimoes commented:
I tried removing the domain fabianaupgrades20151120125457.net from my account, but that left the Domain Mapping associated to it behind: 

![screenshot-calypso localhost 3000 2016-01-22 12-25-05](https://cloud.githubusercontent.com/assets/1248436/12509500/352bcd3a-c103-11e5-8109-81f764a74cec.png)

fabianapsimoes commented:
Right now, removing a plan will leave the domain associated to it behind. That can be confusing to users looking to get rid of all their purchases with WordPress.com. We should add some messaging explaining the domain will remain in the account. cc @mikeshelton1503

fabianapsimoes commented:
Related to the issue I reported [here](https://github.com/Automattic/wp-calypso/pull/2602#issuecomment-173887662), I noticed that removing a domain does **not** cancel its registration. I've tested this with both testingupgrades.org and fabianaupgrades20151120125457.net, and the registrations should have been canceled with WWD.

fabianapsimoes commented:
@Lochlaer I'm curious how well this flow works for Google Apps. Could you give it a try?

You should see:

- "Cancel and Refund" within the refund window, which will remove the subscriptions and trigger a refund.
- "Cancel" outside of the refund window, which will just disable auto-renewal and leave the subscriptions behind, still active until expiration.
- "Remove" also outside the refund window after clicking "Cancel" (it might take a few seconds for the option to appear). This will remove the subscriptions and should terminate Google Apps use.

Seems everything works as expected on WordPress.com. Is it all OK from Google's end as well?

stephanethomas commented:
I have updated the dialog to display a custom text for domain registrations. I have also fixed the heading on small screens. I'm not a big fan of this fix but otherwise it would mean updating styles from the `<Dialog />` shared component - something that is always tricky (note we used a [similar approach](https://github.com/Automattic/wp-calypso/blob/a8ac0e2fcc8c9f1b5fd5a407673f4c697981a88c/assets/stylesheets/sections/_checkout.scss#L628) in the checkout).

stephanethomas commented:
> Right now, removing a plan will leave the domain associated to it behind.

Maybe this is something we just need to fix in the backend (i.e. unbundle the domain)?

fabianapsimoes commented:
> Maybe this is something we just need to fix in the backend (i.e. unbundle the domain)?

We do unbundle it, leaving the domain in the user's account. I think that's a good default, but we should let the user know the domain is still active, in case they were hoping that canceling the plan would also cancel the domain.

drewblaisdell commented:
The endpoint now:

- Cancels domains through the registrar's API when removing domains.
- Removes the domain mapping/private registration subscriptions when removing a domain registration.
- Returns an error when the user attempts to remove a domain that is included with a plan. This is also not possible in this PR.

> I have updated the dialog to display a custom text for domain registrations.

For reference, here is that message:

![screen shot 2016-01-19 at 4 59 10 pm](https://cloud.githubusercontent.com/assets/1130674/12526211/dfdae9d4-c11f-11e5-8445-33e3fa89aae5.png)

I added a commit to update the success notice when removing a domain from this:

![screen shot 2016-01-22 at 2 08 38 pm](https://cloud.githubusercontent.com/assets/1130674/12526194/b01cb452-c11f-11e5-8bdc-ccd6ed5052c4.png)

To this:

![screen shot 2016-01-22 at 3 46 57 pm](https://cloud.githubusercontent.com/assets/1130674/12526198/b688b4e4-c11f-11e5-94f6-aea00f4a4879.png)

It might still need tweaking, though. Also, I'm labeling this as 'Needs Design Review' to determine how to change the messaging when removing a plan that includes a domain.

fabianapsimoes commented:
Still not ideal layout for longer domains in narrow screens:

![screenshot-calypso localhost 3000 2016-01-25 13-18-47](https://cloud.githubusercontent.com/assets/1248436/12550921/04de8f8a-c367-11e5-9a2c-20f36b589510.png)

fabianapsimoes commented:
I have a weird situation here. I just purchased a Premium plan (so it's within the refund window) and then disabled auto-renewal from the admin tools. I now see both "Cancel and Refund" and "Remove" in `/purchases`:

![screenshot-calypso localhost 3000 2016-01-25 14-13-48](https://cloud.githubusercontent.com/assets/1248436/12551868/f0c08bdc-c36d-11e5-8871-c88036459044.png)

I know this is an artificial scenario, but I would have expected these buttons to be displayed based on the same dates? I'm guessing the refund button is looking at the transaction date, and remove is looking at the purchase date?

stephanethomas commented on client/lib/wpcom-undocumented/lib/me.js:
```diff
@@ -285,4 +285,12 @@ UndocumentedMe.prototype.deleteAccountRecoveryEmail = function( callback ) {
 	return this.wpcom.req.post( args, callback );
 };
 
+UndocumentedMe.prototype.removePurchase = function( purchaseId, fn ) {
```
I wonder if we shouldn't name this function after the API endpoint, i.e. `deletePurchase()`.

stephanethomas commented on client/me/purchases/remove-purchase/index.jsx:
```diff
@@ -0,0 +1,181 @@
+			if ( success ) {
+				if ( isDomainRegistration( purchase ) ) {
+					notices.success(
+						this.translate( `The domain {{em}}%(domain)s{{/em}} was removed from {{em}}%(wpcomUrl)s{{/em}}.`, {
```
We don't need string templates for these success notices.

stephanethomas commented on client/me/purchases/remove-purchase/index.jsx:
```diff
@@ -0,0 +1,181 @@
+			<p>
+				{
+					this.translate( 'Are you sure you want to remove %(productName)s from {{em}}%(domain)s{{/em}}?', {
+						args: { productName, domain: this.props.selectedSite.slug },
```
We should probably be consistent and use the same variable names throughout this component. We're using `siteSlug` for the success notice but here we use `domain`.

stephanethomas commented on client/me/purchases/remove-purchase/index.jsx:
```diff
@@ -0,0 +1,181 @@
+						this.translate( `The domain {{em}}%(domain)s{{/em}} was removed from {{em}}%(wpcomUrl)s{{/em}}.`, {
+							args: {
+								domain: purchase.meta,
+								wpcomUrl: selectedSite.wpcom_url
```
I'm not sure we need to specify that url here since the dialog for domains will show the following text:

> This will remove %(domain)s from your account. By removing, you are canceling the domain registration. This may stop you from using it again, even with another service.

I guess we should just mention _account_ again in this success message:

> The domain {{em}}%(domain)s{{/em}} was removed from your account.

Let's see what @mikeshelton1503 or @breezyskies think of that, especially since they will revisit this message for bundle domains anyway.

breezyskies replied:
+1 for the "from your account" copy change.

fabianapsimoes commented:
I've observed the bug described in #2746 while testing this PR. I don't think it relates to any of the changes we're introducing here, so I opened a separate issue.

stephanethomas commented:
I've fixed the layout issue plus most of the things mentioned in my comments.

stephanethomas commented:
> I know this is an artificial scenario, but I would have expected these buttons to be displayed based on the same dates? I'm guessing the refund button is looking at the transaction date, and remove is looking at the purchase date?

@fabianapsimoes, we don't look at dates in Calypso to trigger these links. In the case you mention the remove option is only looking for an expiration status with one of [these](https://github.com/Automattic/wp-calypso/blob/993ccc8d5bc7c391b5f283a5cdaa59597e1a2b7b/client/lib/purchases/index.js#L91) [values](https://github.com/Automattic/wp-calypso/blob/993ccc8d5bc7c391b5f283a5cdaa59597e1a2b7b/client/lib/purchases/index.js#L95):

```
cardExpired
cardExpiring
expired
expiring
manualRenew
```

That indeed includes when auto-renewal has been disabled.

The cancel option checks whether the purchase [is cancelable](https://github.com/Automattic/wp-calypso/blob/993ccc8d5bc7c391b5f283a5cdaa59597e1a2b7b/client/lib/purchases/index.js#L75), and as such only looks for an `expired` status in addition to checking if it's refundable and if auto-renew can be disabled.

Investigating this issue lead me to spend quite some time in our [`purchases` functions](https://github.com/Automattic/wp-calypso/blob/993ccc8d5bc7c391b5f283a5cdaa59597e1a2b7b/client/lib/purchases/index.js) and I think we're doing that wrong. We shouldn't have that much logic in `isCancelable()` and `isRemovable()`: this should live in the API. It is bad to the point that the API even returns a [`is_cancelable` property](https://github.com/Automattic/wp-calypso/blob/993ccc8d5bc7c391b5f283a5cdaa59597e1a2b7b/client/lib/purchases/assembler.js#L29) that we don't use at all in Calypso.

I really think we should only handle flags in Calypso, move that logic behind the API, and take advantage of this specific property. It doesn't have to happen in this pull request though. @drewblaisdell, I'd like to get your thoughts on that when you get the chance.

drewblaisdell commented:
> I really think we should only handle flags in Calypso, move that logic behind the API, and take advantage of this specific property.

I agree with this. Part of the reason for this is the disparity between the API and the client about the use of the term 'cancel', but this is something we should address now that purchases are live and we have a better understanding of what that term means.

@stephanethomas Your changes look good to me. If you're cool with my prior commits as well, we can tick the code box here.

stephanethomas commented:
Code looks good to me.

mikeshelton1503 commented:
>Right now, removing a plan will leave the domain associated to it behind. That can be confusing to users looking to get rid of all their purchases with WordPress.com. We should add some messaging explaining the domain will remain in the account.

Ok how about something like this:
>Are you sure you want to remove WordPress.com Premium? You will not be able to reuse it again without purchasing a new subscription. The domain associated with this plan, example.com, will not be removed. It will remain active on your site, unless also removed.

It's long but I'm not sure how much shorter we can make it. I did remove "By removing it, " to shorten it a bit. So we should also remove that from the other version of this message.

drewblaisdell commented:
I rebased against master and added https://github.com/Automattic/wp-calypso/commit/145b349d350a3222713c9e6f3f3474202a8d20b0 to add a specific cancelation message for plans that include domains per @mikeshelton1503's comment:

![screen shot 2016-01-25 at 5 16 25 pm](https://cloud.githubusercontent.com/assets/1130674/12569994/d7bf4676-c388-11e5-9b44-932a6bc3000c.png)

stephanethomas commented:
I added https://github.com/Automattic/wp-calypso/commit/4e058a665d63e696671cb8e5bcc78fb5136cc7ae to address the comments I made earlier. Note I left `By removing it` for domain registrations since I felt the copy was a bit off without it.

fabianapsimoes commented:
Product :+1:
